source/braille.py

```diff
@@ -1540,47 +1565,77 @@ def _set_tether(self, tether):
 	def _get_shouldAutoTether(self):
 		return self.enabled and config.conf["braille"]["autoTether"]
 
-	def setDisplayByName(self, name, isFallback=False):
-		if not name:
-			self.display = None
-			self.displaySize = 0
-			return
-		# See if the user have defined a specific port to connect to
-		if name not in config.conf["braille"]:
-			# No port was set.
-			config.conf["braille"][name] = {"port" : ""}
-		port = config.conf["braille"][name].get("port")
-		# Here we try to keep compatible with old drivers that don't support port setting
-		# or situations where the user hasn't set any port.
+	def setDisplayByName(self, name, isFallback=False, detected=None):
+		if name == AUTO_DISPLAY_NAME:
+			self._enableDetection()
+			return True
+		elif not isFallback and not detected:
+			self._disableDetection()
+
 		kwargs = {}
-		if port:
-			kwargs["port"] = port
+		if detected:
+			kwargs["port"]=detected
+		else:
+			# See if the user have defined a specific port to connect to
+			if name not in config.conf["braille"]:
+				# No port was set.
+				config.conf["braille"][name] = {"port" : ""}
+			port = config.conf["braille"][name].get("port")
+			# Here we try to keep compatible with old drivers that don't support port setting
+			# or situations where the user hasn't set any port.
+			if port:
+				kwargs["port"] = port
+
 		try:
 			newDisplay = _getDisplayDriver(name)
+			if detected and bdDetect._isDebug():
+				log.debug("Possibly detected display '%s'" % newDisplay.description)
 			if newDisplay == self.display.__class__:
 				# This is the same driver as was already set, so just re-initialise it.
+				log.debug("Reinitializing %s braille display"%name)
 				self.display.terminate()
 				newDisplay = self.display
-				newDisplay.__init__(**kwargs)
+				try:
+					newDisplay.__init__(**kwargs)
+				except TypeError:
+					# Re-initialize with supported kwargs.
+					extensionPoints.callWithSupportedKwargs(newDisplay.__init__, **kwargs)
 			else:
-				if newDisplay.isThreadSafe:
+				if newDisplay.isThreadSafe and not detected:
 					# Start the thread if it wasn't already.
+					# Auto detection implies the thread is already started.
 					_BgThread.start()
-				newDisplay = newDisplay(**kwargs)
+				try:
+					newDisplay = newDisplay(**kwargs)
+				except TypeError:
+					newDisplay = newDisplay.__new__(newDisplay)
+					# initialize with supported kwargs.
+					extensionPoints.callWithSupportedKwargs(newDisplay.__init__, **kwargs)
 				if self.display:
+					log.debug("Switching braille display from %s to %s"%(self.display.name,name))
 					try:
 						self.display.terminate()
 					except:
 						log.error("Error terminating previous display driver", exc_info=True)
 				self.display = newDisplay
 			self.displaySize = newDisplay.numCells
 			self.enabled = bool(self.displaySize)
-			if not isFallback:
+			if isFallback:
+				self._resumeDetection()
+			elif not isFallback and not detected:

```

Comment:
`not isFallback` is redundant since this is the else-statemnet for `isFallback is True`.